### PR TITLE
Review unit test coverage of logic vote - Closes #1663

### LINKED ISSUE #1663: Review unit test coverage of logic/vote
Parent #1302

Expected behavior
Ensure to add missing tests.

Actual behavior
There are some missing tests.

Which version(s) does this affect? (Environment, OS, etc...)
1.0.0

### PR DESCRIPTION
**Pull request by ManuGowda**
### What was the problem?
Missing tests for `getBytes`
### How did I fix it?
Added tests
### How to test it?
`mocha test/unit/logic/vote.js`
### Review checklist

* The PR solves #1663 
* All new code is covered with unit tests
* All new code was formatted with Prettier
* Linting passes
* Tests pass
* Commit messages follow the [commit guidelines](CONTRIBUTING.md#git-commit-messages)
* Documentation has been added/updated

### CODE REVIEW

File: test/unit/logic/vote.js [hunk 1]
```diff
@@ -522,6 +522,40 @@ describe('vote', () => {
 		});
 	});
 
+	describe('getBytes', () => {
+		it('should return null for empty asset', () => {
```
Comment (Tosch110) [inline test/unit/logic/vote.js]: How much sense does a vote asset have without votes?

An empty asset should throw imo

File: test/unit/logic/vote.js [hunk 2]
```diff
@@ -522,6 +522,40 @@ describe('vote', () => {
...
+			var transaction = _.cloneDeep(validTransaction);
+			var data = {
+				votes: [
+					'-9d3058175acab969f41ad9b86f7a2926c74258670fe56b37c429c01fca9f2f0f',
+				],
+			};
+			transaction.asset = data;
+
+			return expect(vote.getBytes(transaction)).to.eql(
+				Buffer.from(data.votes[0], 'utf8')
```
Comment (Tosch110) [inline test/unit/logic/vote.js]: prefer to have this in a const, preferred as direct Buffer or string conversion to Buffer

File: test/unit/logic/vote.js [hunk 3]
```diff
@@ -522,6 +522,40 @@ describe('vote', () => {
...
+			var transaction = _.cloneDeep(validTransaction);
+			transaction.asset = {};
+			return expect(vote.getBytes(transaction)).to.eql(null);
+		});
+
+		it('should return bytes of data asset', () => {
+			var transaction = _.cloneDeep(validTransaction);
+			var data = {
+				votes: [
+					'-9d3058175acab969f41ad9b86f7a2926c74258670fe56b37c429c01fca9f2f0f',
```
Comment (Tosch110) [inline test/unit/logic/vote.js]: please also add a case with a `+publicKey`

File: test/unit/logic/vote.js [hunk 4]
```diff
@@ -522,6 +522,40 @@ describe('vote', () => {
...
+
+			return expect(vote.getBytes(transaction)).to.eql(
+				Buffer.from(data.votes[0], 'utf8')
+			);
+		});
+
+		it('should be okay for utf-8 data value', () => {
+			var transaction = _.cloneDeep(validTransaction);
+			var data = {
+				votes: ['Zu¨¨¨¨ka Ωlaå'],
```
Comment (Tosch110) [inline test/unit/logic/vote.js]: votes is always an array with minus or plus appended public keys, it makes more sense to me to have actually this pattern in the test. IMO every other case can throw an error.

File: test/unit/logic/vote.js (hunk 1 above)
Comment (MaciejBaj) [reply test/unit/logic/vote.js]: Good point.

File: test/unit/logic/vote.js [hunk 5]
```diff
@@ -522,6 +522,75 @@ describe('vote', () => {
 		});
 	});
 
+	describe('getBytes', () => {
+		it('should throw error for empty votes', () => {
+			var transaction = _.cloneDeep(validTransaction);
+			transaction.asset = {};
+			return expect(() => {
+				vote.getBytes(transaction);
+			}).to.throw();
```
Comment (Tosch110) [inline test/unit/logic/vote.js]: Adding the error message that is thrown would make sense here

File: test/unit/logic/vote.js [hunk 6]
```diff
@@ -522,6 +522,75 @@ describe('vote', () => {
...
+		});
+
+		it('should throw error for zero votes', () => {
+			var transaction = _.cloneDeep(validTransaction);
+			transaction.asset.votes = [];
+			return expect(() => {
+				vote.getBytes(transaction);
+			}).to.throw();
```
Comment (Tosch110) [inline test/unit/logic/vote.js]: Adding the error message would be good, to know the right error is triggered

File: test/unit/logic/vote.js [hunk 7]
```diff
@@ -522,6 +522,75 @@ describe('vote', () => {
...
+			var transaction = _.cloneDeep(validTransaction);
+			const votes = new Array(34);
+			_.fill(
+				votes,
+				'-9d3058175acab969f41ad9b86f7a2926c74258670fe56b37c429c01fca9f2f0f'
+			);
+			transaction.asset.votes = votes;
+			return expect(() => {
+				vote.getBytes(transaction);
+			}).to.throw();
```
Comment (Tosch110) [inline test/unit/logic/vote.js]: missing error message

File: test/unit/logic/vote.js [hunk 8]
```diff
@@ -522,6 +522,75 @@ describe('vote', () => {
...
+		});
+
+		it('should throw error for votes with invalid public key', () => {
+			var transaction = _.cloneDeep(validTransaction);
+			transaction.asset.votes = [
+				'-9d3058175acab969f41ad9b86f7a2926c74258670fe56b',
+			];
+			return expect(() => {
+				vote.getBytes(transaction);
+			}).to.throw();
```
Comment (Tosch110) [inline test/unit/logic/vote.js]: error message

File: test/unit/api/ws/middleware/emit.js [hunk 9]
```diff
@@ -51,7 +51,11 @@ describe('emitMiddleware', () => {
 		});
 
 		describe('when req.event is one of the events responsible for receiving data on P2P layer', () => {
-			const receiveDataEvents = ['postBlock', 'postTransactions', 'postSignatures'];
+			const receiveDataEvents = [
```
[the remark this answers is not in the document]
Comment (ManuGowda) [reply test/unit/api/ws/middleware/emit.js]: ahh, this is not relevant to my changes, wired prettier.

File: test/unit/logic/vote.js [hunk 10]
```diff
@@ -522,6 +522,94 @@ describe('vote', () => {
...
+			var transaction = _.cloneDeep(validTransaction);
+			transaction.asset.votes = [
+				'$9d3058175acab969f41ad9b86f7a2926c74258670fe56b37c429c01fca9f2f0f',
+			];
+			const prefix = transaction.asset.votes[0].charAt(0);
+			const err = `Expected a prefix '+' for vote and '-' for unvote in public key, but found ${prefix}`;
+
+			return expect(() => {
+				vote.getBytes(transaction);
+			}).to.throw(err);
```
Comment (Tosch110) [inline test/unit/logic/vote.js]: Can you update the other tests to have the same pattern, the error message in a declared string and point to it in the assertion?

Comment (ManuGowda) [reply test/unit/logic/vote.js]: the getBytes is not used in the `logic/votes` functions so I don't see any other place where error being thrown, or am I missing something?

Comment (Tosch110) [reply test/unit/logic/vote.js]: Your approach works fine :+1:

**Comment by MaciejBaj** [comment]
Thanks Tobi @Tosch110

**Comment by ManuGowda** [comment]
@MaciejBaj the votes were not being validated at first place and @Tosch110 requested these changes which I think is valid, but @Tosch110 you can comment on it, and let me know what's the better approach for this PR.

File: helpers/constants.js [hunk 11]
```diff
@@ -135,6 +135,7 @@ var constants = {
 	// WARNING: When changing totalAmount you also need to change getBlockRewards(int) SQL function!
 	totalAmount: 10000000000000000,
 	unconfirmedTransactionTimeOut: 10800, // 1080 blocks
+	PUBLIC_KEY_HEX_LENGTH: 64,
```
Comment (karmacoma) [inline helpers/constants.js]: Imo, this should be named `publicKeyHexLength` according to other properties in this object. Deviating from this notation style creates an inconsistency.

File: logic/vote.js [hunk 12]
```diff
@@ -310,9 +310,35 @@ Vote.prototype.process = function(transaction, sender, cb) {
  */
 Vote.prototype.getBytes = function(transaction) {
 	try {
-		return transaction.asset.votes
-			? Buffer.from(transaction.asset.votes.join(''), 'utf8')
-			: null;
+		const votes = transaction.asset.votes;
+		if (!Array.isArray(votes)) {
+			throw new Error('votes parameter must be an Array.');
```
Comment (karmacoma) [inline logic/vote.js]: Sentence case: `Votes parameter must be an array`

File: logic/vote.js [hunk 13]
```diff
@@ -310,9 +310,35 @@ Vote.prototype.process = function(transaction, sender, cb) {
...
+		}
+
+		if (
+			!(votes.length > 0 && votes.length < constants.maxVotesPerTransaction)
+		) {
+			throw new Error('number of votes must be between 0 and 33.');
```
Comment (karmacoma) [inline logic/vote.js]: Sentence case: `Number of votes must be between 0 and 33`

**Comment by Tosch110** [comment]
I do not quite understand. Maybe there is a common practise you have when the tests reveal that the code is not complete? But I would say that when adding test coverage and you identify anything in the tests, the code should be updated accordingly to make the tests pass.

**Comment by ManuGowda** [comment]
@Tosch110 I think the explanation to why `getBytes` function doesn't have checks is that here [https://github.com/LiskHQ/lisk/blob/c520928284262b582cce54f80ee061d6f2b3e56b/modules/blocks/verify.js#L288](url) we are checking the transaction. so if we add checks in each `getBytes` it would be redundant, @4miners correct me if I am wrong.

**Comment by MaciejBaj** [comment]
The `getByte` function in Vote the rest of transaction modules does what it says - gets bytes. Currently for the verification of the transaction's content is done in `verify` step of all transaction types. I don't see any reason to move it / duplicate it in getBytes functions of all transactions.

**Pull request merged**